### PR TITLE
+per #17894 Support Future(Nil) in asyncWriteMessages

### PR DESCRIPTION
**Pull request by patriknw**
Future.successful(Nil) is an alternative way to signal all good
in the happy case, for reduced allocations.

### CODE REVIEW

**Comment by patriknw** [comment]
Refs #17894

File: akka-persistence/src/main/scala/akka/persistence/journal/leveldb/LeveldbStore.scala [hunk 1]
```diff
@@ -42,9 +42,24 @@ private[persistence] trait LeveldbStore extends Actor with WriteJournalBase with
...
+          } catch {
+            case NonFatal(e) ⇒
+              if (result.isEmpty)
+                result = Vector.fill(n)(AsyncWriteJournal.successUnit)
+              result :+= Failure(e)
+          }
+        }
+
+        result
+      }
```
Comment (patriknw) [inline akka-persistence/src/main/scala/akka/persistence/journal/leveldb/LeveldbStore.scala]: I'm not sure we should do this change in the leveldb impl. It's so much uglier.

Comment (ktoso) [inline akka-persistence/src/main/scala/akka/persistence/journal/leveldb/LeveldbStore.scala]: I wonder how much perf impact it has, it's quite a few allocations less.. hm

Comment (drewhk) [inline akka-persistence/src/main/scala/akka/persistence/journal/leveldb/LeveldbStore.scala]: what was the reason for the change?

Comment (patriknw) [inline akka-persistence/src/main/scala/akka/persistence/journal/leveldb/LeveldbStore.scala]: as far as I can see the old impl allocates one Success(Unit) instance per message, while the new impl only allocates one iterator (for the happy path)

Comment (patriknw) [inline akka-persistence/src/main/scala/akka/persistence/journal/leveldb/LeveldbStore.scala]: The purpose with the PR is to support Future(Nil) in asyncWriteMessages so that implementors of journal plugins can use that for optimizations (reduce allocations). It was requested by some (and we also discussed it early).

Comment (patriknw) [inline akka-persistence/src/main/scala/akka/persistence/journal/leveldb/LeveldbStore.scala]: We don't have to change the leveldb impl, and I don't think we should (we don't care about that optimization)

**Comment by patriknw** [comment]
the build failure is real, I'll fix

File: akka-persistence/src/main/scala/akka/persistence/journal/leveldb/LeveldbStore.scala (hunk 1 above)
Comment (patriknw) [inline akka-persistence/src/main/scala/akka/persistence/journal/leveldb/LeveldbStore.scala]: I have fixed the test failure, and it was actually a [bug in my implementation here](https://github.com/akka/akka/commit/2fc091c33ffc6bd286f3ed9f743b69794a62c81c). Found by tck.

I vote for not changing leveldb impl. Ok @ktoso ?

Comment (ktoso) [inline akka-persistence/src/main/scala/akka/persistence/journal/leveldb/LeveldbStore.scala]: We can leave it out from this PR. We don't recommend LevelDB for anything other than toy projects anyway...

I'm curious how it impacted perf of happy case (raw throughput with persistAsync), since we have it implemented could you run the `PersistentActorThroughputBenchmark` with it before we revert the change?

Comment (patriknw) [inline akka-persistence/src/main/scala/akka/persistence/journal/leveldb/LeveldbStore.scala]: will do

Comment (patriknw) [inline akka-persistence/src/main/scala/akka/persistence/journal/leveldb/LeveldbStore.scala]: you didn't say that it would take 2 x 36 minutes ;-)
I'll do the second run when having dinner or something.

Here is the first run, with the optimized code:

    [info] Benchmark                                                                                    Mode  Cnt       Score      Error  Units
    [info] PersistentActorThroughputBenchmark.actor_normalActor_reply_baseline                         thrpt  200  146964.946 ± 2766.057  ops/s
    [info] PersistentActorThroughputBenchmark.persistentActor_noPersist_reply                          thrpt  200  149247.984 ± 2760.938  ops/s
    [info] PersistentActorThroughputBenchmark.persistentActor_persistAsync_reply                       thrpt  200  107536.480 ± 2585.782  ops/s
    [info] PersistentActorThroughputBenchmark.persistentActor_persistAsync_replyRightOnCommandReceive  thrpt  200  110368.506 ± 2786.160  ops/s
    [info] PersistentActorThroughputBenchmark.persistentActor_persist_reply                            thrpt  200    9601.518 ±   81.373  ops/s

Comment (ktoso) [inline akka-persistence/src/main/scala/akka/persistence/journal/leveldb/LeveldbStore.scala]: lol sorry, that's a lot of runs :d
Looking forward to it :)

// me writing blogpost..

Comment (patriknw) [inline akka-persistence/src/main/scala/akka/persistence/journal/leveldb/LeveldbStore.scala]: here is the second run, with the change here reverted:

    [info] PersistentActorThroughputBenchmark.actor_normalActor_reply_baseline                         thrpt  200  143879.774 ± 2125.070  ops/s
    [info] PersistentActorThroughputBenchmark.persistentActor_noPersist_reply                          thrpt  200  150487.166 ± 2621.352  ops/s
    [info] PersistentActorThroughputBenchmark.persistentActor_persistAsync_reply                       thrpt  200  111451.952 ± 2536.184  ops/s
    [info] PersistentActorThroughputBenchmark.persistentActor_persistAsync_replyRightOnCommandReceive  thrpt  200  114634.707 ± 2604.804  ops/s
    [info] PersistentActorThroughputBenchmark.persistentActor_persist_reply                            thrpt  200    9686.627 ±   36.865  ops/s 

i.e. no difference

Comment (ktoso) [inline akka-persistence/src/main/scala/akka/persistence/journal/leveldb/LeveldbStore.scala]: Interesting... thanks for checking! It's not dominating the work in any case but good to see it's practically no impact.

Ok to remove this optimisation from our Journal then :)

**Comment by patriknw** [comment]
reverted the leveldb change, rebased

**Comment by ktoso** [comment]
LGTM
